[NET-316][system-probe] Count DNS responses received for each connection

mbakht opened the pull request:
### What does this PR do?

This PR adds a metric to count successful DNS responses. 
JIRA Ticket: https://datadoghq.atlassian.net/browse/NET-316

### Motivation

What inspired you to submit this pull request?

### Testing

A small python script `parser.py`:

```python
import json
import sys

d = json.load(sys.stdin)
for c in d['conns']:
    print "{: <5} {} {: >12}:{: <5} ->{: >16}:{: <5} {: >2}".format(c['pid'], c['type'], c['laddr']['ip'], c['laddr']['port'], c['raddr']['ip'], c['raddr']['port'], c['dnsSuccessfulResponses'])
```

```
vagrant@agent-dev-mehedi:~/go/src/github.com/DataDog/datadog-agent$ curl -XGET --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections | python parser.py
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   800  100   800    0     0   111k      0 --:--:-- --:--:-- --:--:--  111k
9550  tcp    10.0.2.15:22    ->        10.0.2.2:56614  0
8682  udp    127.0.0.1:52152 ->       127.0.0.1:8125   0
```

The last column shows successful DNS responses.

Now, lets do a `wget`. For example:

```
vagrant@agent-dev-mehedi:~/go/src/github.com/DataDog/datadog-agent$ wget golang.org
```

Now if we run the previous command, there should be some nonzero values in the right most column:

```
vagrant@agent-dev-mehedi:~/go/src/github.com/DataDog/datadog-agent$ curl -XGET --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections | python parser.py
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2822    0  2822    0     0   196k      0 --:--:-- --:--:-- --:--:--  196k
8858  tcp    10.0.2.15:56496 ->    172.217.9.49:443    0
8858  tcp    10.0.2.15:43220 ->    172.217.9.49:80     0
8858  udp    127.0.0.1:34225 ->      127.0.0.53:53     1
827   udp    10.0.2.15:46109 ->        10.0.2.3:53     1
9550  tcp    10.0.2.15:22    ->        10.0.2.2:56614  0
8682  udp    127.0.0.1:52152 ->       127.0.0.1:8125   0
827   udp    10.0.2.15:59970 ->        10.0.2.3:53     0
```

mbakht committed:
Initial commit

mbakht committed:
Merge branch 'master' into mehedi.bakht/dns-replies

mbakht committed:
Add test

mbakht committed:
tests work

mbakht committed:
Clean up code

mbakht committed:
code reformat

kevinconaway commented on pkg/ebpf/dns_snooper_test.go:
```diff
@@ -4,6 +4,7 @@ package ebpf
 
 import (
 	bpflib "github.com/iovisor/gobpf/elf"
+	"k8s.io/apimachinery/pkg/util/rand"
```
nit, did you mean to import `math/rand` here?

mbakht replied:
🤦‍♂

kevinconaway commented on pkg/ebpf/dns_stats.go:
```diff
@@ -0,0 +1,47 @@
+	d.mux.Lock()
+	defer d.mux.Unlock()
+	stats := d.stats[key]
+	if stats.lastTransactionID == transactionID {
```
Are there any ordering concerns here with how packets arrive that could make this incorrect?

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -406,6 +406,18 @@ func (t *Tracer) getConnections(active []ConnectionStats) ([]ConnectionStats, ui
+						queryPort: conn.SPort,
+						protocol:  conn.Type,
+					}
+					dnsStats := t.reverseDNS.GetDNSStats(key)
```
Any reason this can't just take a conn and have it's internal key representation constructed within `GetDNSStats()`?

mbakht committed:
Use updated proto file

mbakht committed:
Update state.go to keep track of DNS stats

mbakht committed:
Fix import

mbakht commented:
@DataDog/networks, this is still a work-in-progress. I am working on adding more tests and addressing comments that have been already made. But it would be good to have some feedback on the current approach, especially how the DNS stats are being managed in `NetworkState`.

mbakht committed:
Merge branch 'master' into mehedi.bakht/dns-replies

mbakht committed:
add DNS stats in a separate function

mbakht committed:
keep track of collisions

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -318,6 +323,15 @@ func (t *Tracer) GetActiveConnections(clientID string) (*Connections, error) {
+			continue
+		}
+		t.state.StoreDNSStats(key, dnsStats)
+	}
```
Can we move the calls to StoreDNSStats above the calls to Connections and change Connections to do all the DNS adding the AddDNSStats is doing ?

mbakht replied:
I initially had it exactly that way. The problem was that for new clients, there was no local `dnsStats` to use in the code block that we have for `AddDNSStats`. There are a few options I can consider:
1. Ignore this since it happens only on the first call for a new client
2. Instead `StoreDNSStats`, we also pass a `clientID` and then in `StoreDNSStats`, we create one if it does not exist (just like what we do in `Connections`)
3.  Have a global copy of `DNSStats` in `NetworkState` that new clients can fall back to.

Please suggest if I am missing something obvious.

leeavital replied:
Could we have `Connections()` take the most recent  copy of the DNS stats and copy them into each client state (as well as initialize any clientState that doesn't exist)?

leeavital commented on pkg/ebpf/state.go:
```diff
@@ -160,6 +174,41 @@ func (ns *networkState) Connections(id string, latestTime uint64, latestConns []
+		if dnsStats, ok := ns.clients[id].dnsStats[key]; ok {
+			conn.DNSReplyCount = dnsStats.replies
+		}
+		seen[key] = struct{}{}
```
can you delete from dnsStats instead of keeping a map for seen connections?

mbakht replied:
Sure. The only reason for doing this way was to add code in the next iteration to detect collisions and probably update some metric (like we do for pid collisions when adding TCP stats). But if it doesn't sound like a metric that we would be interested in, I can directly delete from `dnsStats` as you suggested.

leeavital commented on pkg/ebpf/dns_stats.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+type dnsStatKeeper struct {
+	mux  sync.Mutex
```
I think we should try to avoid using a mutex if possible.

kevinconaway commented on pkg/ebpf/dns_stats.go:
```diff
@@ -0,0 +1,62 @@
+
+	// For local DNS traffic, sometimes the same reply packet gets processed by the
+	// snooper multiple times. This check avoids double counting in that scenario.
+	if dnsData.lastTransactionID == transactionID {
```
If the same packet can be processed more than once, is it possible to for other replies to be interleaved in? For example, `p1, p2, p1, p1` ?

mbakht replied:
One possible reason for duplication of the local DNS traffic is the [following:](https://stackoverflow.com/questions/9196791/duplicate-udp-packets-how-often-it-happens) 

> If the socket in the destination host is bound to INADDR_ANY and if the source broadcasts the UDP packets then there is possibility to receive duplicate packets

I guess in that scenario it is technically possible for packets to be interleaved though I have not been able to reproduce it. To take care of that possibility, we can keep a set of last N transaction IDs and check against that set.
However, this duplication happens only for local DNS traffic which we discard by default. In the rare case that we enable it, checking against the last transaction ID would take care of most if not all of the cases. So the code ignores the highly unlikely case of interleaved duplicate packets in exchange for faster checks in the common case. I will add a comment to that effect. 

If the tradeoff does not make sense or if there is a way to avoid the tradeoff, I would be happy to listen to other ideas.

kevinconaway replied:
I think thats fine, thanks for explaining.

mbakht committed:
Get rid of extra copying. Pass local DNS config to snooper

mbakht committed:
Use latest version of agent-payload

mbakht committed:
Merge branch 'master' into mehedi.bakht/dns-replies

kevinconaway commented on pkg/ebpf/event_common.go:
```diff
@@ -107,6 +107,7 @@ type ConnectionStats struct {
 	Direction     ConnectionDirection
 	IPTranslation *netlink.IPTranslation
 	IntraHost     bool
+	DNSReplyCount uint32
```
Do we want to rename this (and elsewhere) to successful replies?

kevinconaway commented on pkg/ebpf/state.go:
```diff
@@ -156,10 +178,41 @@ func (ns *networkState) Connections(id string, latestTime uint64, latestConns []
+		}
+
+		if _, alreadySeen := seen[key]; alreadySeen {
+			ns.telemetry.dnsPidCollisions++
```
why are we calling this pid collisions?  it looks like it is colliding on the (srcip, dstip, sport, dport) combo?

mbakht replied:
I tried to be consistent with how we deal with a similar situation when processing TCP stats which also does not have PID information: https://github.com/DataDog/datadog-agent/blob/master/pkg/ebpf/tracer.go#L501

I think the idea is that different PIDs are ending up being mapped to the same value - so they are _colliding_ 😕 .

leeavital commented on pkg/ebpf/dns_stats.go:
```diff
@@ -0,0 +1,58 @@
+
+	// There is only one writer which is this thread. So we need to lock only
+	// before writing, not reading.
+	d.mux.Lock()
```
I think this comment might be stale, GetAndResetAllStats() locks too

leeavital reviewed (commented):
This is coming together really nicely! we should add some tests in state_test.go to cover the stat merging.

mbakht committed:
Rename to successful responses, fix data race

mbakht commented:
> This is coming together really nicely! we should add some tests in state_test.go to cover the stat merging.

Sure, I will work on adding some stats. I just wanted to make sure first that overall approach for state management looks good.

mbakht committed:
Add state tests

mbakht committed:
Merge branch 'master' into mehedi.bakht/dns-replies

mbakht committed:
Add comment about outbound IP

mbakht committed:
Merge branch 'master' into mehedi.bakht/dns-replies

mbakht committed:
add release note

mbakht committed:
remove extra blank line

mbakht committed:
Add feature flag

mbakht committed:
Merge branch 'master' into mehedi.bakht/dns-replies
